File: machine/cpp/machine/cpu/source_text.cpp
#include "machine/cpu/source_text.h"

namespace bmsx {

std::optional<std::string> extractSourceRangeText(const SourceRange& range, const std::string& source) {
	if (range.start.line <= 0 || range.end.line < range.start.line) {
		return std::nullopt;
	}
	std::string out;
	int line = 1;
	size_t lineStart = 0;
	for (size_t index = 0; index <= source.size(); ++index) {
		if (index < source.size() && source[index] != '\n') {
			continue;
		}
		size_t lineEnd = index;
		if (lineEnd > lineStart && source[lineEnd - 1] == '\r') {
			lineEnd -= 1;
		}
		if (line >= range.start.line && line <= range.end.line) {
			if (!out.empty()) {
				out.push_back(' ');
			}
			out.append(source.data() + lineStart, lineEnd - lineStart);
			if (line == range.end.line) {
				return out;
			}
		}
		line += 1;
		lineStart = index + 1;
	}
	return std::nullopt;
}

} // namespace bmsx

File: machine/ts/machine/cpu/source_text.ts
import type { SourceRange } from './cpu';

export function extractSourceRangeText(range: SourceRange, source: string): string | null {
	if (range.start.line <= 0 || range.end.line < range.start.line) {
		return null;
	}
	let line = 1;
	let lineStart = 0;
	let out = '';
	for (let index = 0; index <= source.length; index += 1) {
		if (index < source.length && source.charCodeAt(index) !== 10) {
			continue;
		}
		let lineEnd = index;
		if (lineEnd > lineStart && source.charCodeAt(lineEnd - 1) === 13) {
			lineEnd -= 1;
		}
		if (line >= range.start.line && line <= range.end.line) {
			if (out.length !== 0) {
				out += ' ';
			}
			out += source.slice(lineStart, lineEnd);
			if (line === range.end.line) {
				return out;
			}
		}
		line += 1;
		lineStart = index + 1;
	}
	return null;
}
